File: l/sorting_algorithm.lua
function bubbleSort(arr)
    local n = #arr
    for i = 1, n - 1 do
        for j = 1, n - i do
            if arr[j] > arr[j + 1] then
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
            end
        end
    end
end

local arr = {64, 34, 25, 12, 22, 11, 90}
bubbleSort(arr)
print("Sorted array: " .. table.concat(arr, ", "))

File: e/sorting_algorithm.ex
defmodule SortingAlgorithm do
  def bubble_sort(arr) do
    do_bubble_sort(arr, length(arr))
  end

  defp do_bubble_sort(arr, 0), do: arr
  defp do_bubble_sort(arr, n) do
    do_bubble_sort(do_bubble(arr), n - 1)
  end

  defp do_bubble([x]), do: [x]
  defp do_bubble([x, y | tail]) when x > y do
    [y | do_bubble([x | tail])]
  end
  defp do_bubble([x, y | tail]), do: [x | do_bubble([y | tail])]
end

arr = [64, 34, 25, 12, 22, 11, 90]
sorted_arr = SortingAlgorithm.bubble_sort(arr)
IO.puts("Sorted array: #{inspect(sorted_arr)}")
